klib/gcp.c

```diff
@@ -698,6 +701,45 @@ static void gcp_metrics_add_memory(buffer body, boolean first, const char *type,
                   first ? "" : ",", gcp.zone, gcp.instance_id, type, state, value, interval);
 }
 
+static void gcp_metrics_add_disk(buffer body, boolean first, const char *type, const char *state,
+                                   u64 value, const char *interval, const char *device)
+{
+    bprintf(body, "%s{\"resource\":{\"type\":\"gce_instance\""
+                  ",\"labels\":{\"zone\":\"%b\",\"instance_id\":\"%b\"}}"
+                  ",\"metric\":{\"type\":\"agent.googleapis.com/disk/%s\""
+                  ",\"labels\":{\"device\":\"%s\",\"state\":\"%s\"}}"
+                  ",\"points\":[{\"value\":{\"doubleValue\":%ld},\"interval\":%s}]}",
+                  first ? "" : ",", gcp.zone, gcp.instance_id, type, device, state, value, interval);
+}
+
+closure_function(3, 4, void, gcp_metrics_disk_vh,
+                 buffer, b, const char *, interval, boolean, skip_readonly,
+                 u8 *, uuid, const char *, label, filesystem, fs, inode, mount_point)
+{
+    if (filesystem_is_readonly(fs) && bound(skip_readonly))
+        return;
+
+    int block_size = fs_blocksize(fs);
+    u64 total_blocks = fs_totalblocks(fs);
+    u64 free_blocks = fs_freeblocks(fs);
+    u64 percent_used_free = free_blocks * 100 / total_blocks;
+    buffer b = bound(b);
+    const char *interval = bound(interval);
+    gcp_metrics_add_disk(b, false, "bytes_used", "free", free_blocks * block_size, interval, label);
+    gcp_metrics_add_disk(b, false, "bytes_used", "used", (total_blocks - free_blocks) * block_size, interval, label);
+    gcp_metrics_add_disk(b, false, "percent_used", "free", percent_used_free, interval, label);
+    gcp_metrics_add_disk(b, false, "percent_used", "used", 100 - percent_used_free, interval, label);
```

Comment:
A volume may have an empty label; in this case, I would use the volume UUID as device identifier, instead of an empty string. See for example the `mounts_handler` closure in src/unix/special.c

Reply:
Checked that initially, but was not sure about the correct implementation, thanks for the example. Pushed some changes.